### PR TITLE
Automatically provision AWS resources with Terraform

### PR DESCRIPTION
**Pull request by 3m3x**
# Background
Provisioning the infrastructure to run awslambdaproxy is manual, tedious and laborious. This friction presents a barrier to first-time usage.

# Changes
This PR introduces a Terraform script which provisions:
* a VPC and internet gateway
* a subnet and route table
* an EC2 instance for the `awslambdaproxy` program
* a Security Group which permits access to the EC2 instance, and permits traffic between the EC2 instance and the Lambdas
* an IAM EC2 role with the `iam/setup.json` policy attached

A simple wrapper script `deploy.sh` kicks off the deployment and then presents the user with a ssh command to run that will start the proxy. If I weren't so lazy I would've put the proxy under control of the supervisor daemon but I think this is good enough for now.

The "Installation" section of `README.md` has also been updated to include usage of this script.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -92,7 +110,7 @@ aws iam create-access-key --user-name awslambdaproxy-run
 6. <b>How much does this cost?</b> awslambdaproxy should be able to run mostly on the [AWS free tier](https://aws.amazon.com/free/) minus bandwidth costs. It can run on a t2.micro instance and the default 128MB Lambda function that is constantly running should also fall in the free tier usage. The bandwidth is what will cost you money; you will pay for bandwidth usage for both EC2 and Lambda.
 7. <b>Why does my connection drop periodically?</b> AWS Lambda functions can currently only execute for a maximum of 15 minutes. In order to maintain an ongoing proxy a new function is executed and all new traffic is cut over to it. Any ongoing connections to the previous Lambda function will hard stop after a timeout period. You generally won't see any issues for normal web browsing as connections are very short lived, but for any long lived connections you may see issues.
 
-# Powered by
+j Powered by
```
Comment (dan-v) [inline README.md]: ?

File: aws.tf [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+  name = "setup_policy"
+  role = "${aws_iam_role.assume_role.id}"
+
+  # This policy's privileges cover both the setup and runtime cases
+  policy = "${file("${var.iam-policy-file}")}"
+}
+
+resource "aws_instance" "lambdaproxybox" {
+  # ap-southeast-2 18.04 LTS "bionic" HVM EBS store
+  ami           = "ami-0532935b53d8e05ee"
```
Comment (dan-v) [inline aws.tf]: make this a variable

File: aws.tf [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+  instance_type = "t2.micro"
```
Comment (dan-v) [inline aws.tf]: make this a variable

File: aws.tf [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+    public_ip = "${aws_instance.lambdaproxybox.public_ip}"
+  }
+
+  # Terraform's ridiculous default approach to ssh'ing into the
+  # instance with the root account doesn't play nicely with our
+  # Ubuntu AMI, so this:
+  connection {
+    type  = "ssh"
+    host  = "${aws_instance.lambdaproxybox.public_ip}"
+    user  = "ubuntu"
```
Comment (dan-v) [inline aws.tf]: make this a variable

File: deploy.sh [hunk 5]
```diff
@@ -0,0 +1,14 @@
+#!/bin/bash
+
+#terraform init
+#terraform plan -out terraform.plan
+#terraform apply terraform.plan
+#IP=$(terraform output lambdaproxybox-ec2-ip)
+IP=54.123.23.300
```
Comment (dan-v) [inline deploy.sh]: all of the terraform code here is commented out and this is hardcoded, guessing you just forgot to swap it back from testing

File: aws.tf [hunk 6]
```diff
@@ -0,0 +1,161 @@
+variable "aws-profile" {}
+variable "ssh-pub-key" {
+  default = "~/.ssh/id_rsa.pub"
+}
+variable "iam-policy-file" {
+  default = "./iam/setup.json"
+}
+variable "package-url" {
```
Comment (dan-v) [inline aws.tf]: make version of package a variable

**Review by dan-v — CHANGES_REQUESTED** [review]
thanks for the PR @3m3x! i definitely like the overall idea of simplifying the setup with terraform. there's a few things i'd like to see tweaked before accepting:
* rather than setting defaults in the tf file (like regions, ssh key, etc), if you could pull that out into a config file it would be easier to see what options are being set / can be customized. the readme could just have an additional bit about tweaking the config file to your liking or whatever.
* for security reasons, i'd definitely prefer the primary port 8080 was not open to the world by default. if this could be locked down to users current IP by default that would be better (could still have option to open to world if that's what someone wants).
* for security reasons, if you could generate a random password or have it be a config value that would be better as well. this is especially important if it is being opened to the world as it would have default well known values right now.
* if you could update the existing deploy script to have option to tear down the terraform infrastructure or create a separate script to do that it would help make this a bit more streamlined. the current approach with remote-exec is kind of a one shot deal on deployment, so if you want to tweak or make changes it would be easy enough to tear it down, edit config file, and deploy a new one.
* agreed that it would be nice to have it run in supervisor like systemd, but fine with the way it is.

File: README.md (hunk 1 above)
Comment (3m3x) [reply README.md]: sorry, typo

**Comment by 3m3x** [comment]
These changes address the points you raised earlier.

The proxy is only accessible via a SSH tunnel that is locally bound to 127.0.0.1:8080 which also makes configuring the browser's proxy settings much simpler.

TODO:
* update README

**Review by dan-v — COMMENTED** [review]
This is looking much better! Thanks for taking the time to fix these items. Once you update the README LGTM.

**Comment by dan-v** [comment]
Closing this out in favor of #15 as it's more comprehensive. Thanks for the initial PR though @3m3x!

**Pull request closed without merging**